File: 4week/1task/Basic/Complex/Complex.cpp
#include <iostream>
#include "Complex.h"

Complex::Complex(double real, double imaginary) {
    this->real = real;
    this->imaginary = imaginary; 
}

Complex::~Complex() { }

double Complex::getReal() const {
    return this->real;
}

double Complex::getImaginary() const {
    return this->imaginary;
}

bool Complex::operator==(const Complex& other) const {
    return this->real == other.real && this->imaginary == other.imaginary;
}

bool Complex::operator!=(const Complex& other) const {
    return this->real != other.real || this->imaginary != other.imaginary;
}

void Complex::operator+=(const Complex& other) {
    this->real += other.real;
    this->imaginary += other.imaginary;
}

void Complex::operator-=(const Complex& other) {
    this->real -= other.real;
    this->imaginary -= other.imaginary;
}

Complex Complex::operator+(const Complex& other) const {
    Complex result = *this;
    result.imaginary += other.imaginary;
    result.real += other.real;
    return result;
}

Complex Complex::operator-(const Complex& other) const {
    Complex result = *this;
    result.imaginary -= other.imaginary;
    result.real -= other.real;
    return result;
}

Complex Complex::operator*(const Complex& other) const {
    Complex result = *this;
    result.imaginary *= other.imaginary;
    result.real *= other.real;
    return result;
}

std::ostream& operator<<(std::ostream& out, const Complex& complex) {
    if ( complex.getImaginary() < 0 ) {
        return out << complex.getReal() << complex.getImaginary();
    } else {
        return out << complex.getReal() << '+' << complex.getImaginary();
    }
}

File: 4week/3task/Complex/Complex.java
public class Complex {
    private double real;
    private double imaginary;
    
public Complex(double real, double imaginary) {
        this.real = real;
        this.imaginary = imaginary;
}

public Complex() {
        this(0,0);
}

public double getReal() {
        return this.real;
}

public double getImaginary() {
        return this.imaginary;
}

public boolean equal(Complex other) {
        return this.real == other.real && this.imaginary == other.imaginary;
}

public void increment(Complex other){
    this.real += other.real;
    this.imaginary += other.imaginary;
}

public Complex sum(Complex other) {
    Complex a = this;
    a.increment(other);
        return a;
}

public void decrement(Complex other) {
    this.real -= other.real;
    this.imaginary -= other.imaginary;
}

public Complex diff(Complex other) {
    Complex a = this;
    a.decrement(other);
        return a;
}

public Complex multiply(Complex other) {
        Complex multiply = this;

        multiply.real *= other.real;
        multiply.imaginary *= other.imaginary;

        return multiply;
}

public void print() {
    System.out.println("("+this.real+";"+this.imaginary+"i"+")");
}

}